Completed Stage 2

vdmppv-xbsoftware opened the pull request.

vdmppv-xbsoftware committed:
Activities section in Contacts executed

vdmppv-xbsoftware committed:
Files section in Contacts executed

vdmppv-xbsoftware committed:
Contacts form executed

vdmppv-xbsoftware committed:
fixes

vdmppv-xbsoftware committed:
fixed labelWidth

vdmppv-xbsoftware committed:
fixed css

vdmppv-xbsoftware committed:
url/selection fix

vdmppv-xbsoftware committed:
added list item photo

kullias commented on sources/views/contactsView.js:
```diff
@@ -1,12 +1,71 @@
+			type: {
+				height: 50,
+				css: "contacts-list-item",
+				template: `<img class="list-item-pic" src="#Photo#"></img> 
```
![Selection_999(860)](https://user-images.githubusercontent.com/31205423/166439443-480c894a-2373-42d1-8780-e6527ce98c37.png)
стоит предусматривать длинные имена и чикать их через ellipsis

kullias commented on sources/views/contactsView.js:
```diff
@@ -1,12 +1,71 @@
+
+		this.on(this.list, "onAfterSelect", (id) => {
+			this.show(`contactsInfo?id=${id}`);
+		});
```
странновато, когда селект с листа не снимается, если добавляешь новый айтем
![Selection_999(862)](https://user-images.githubusercontent.com/31205423/166440441-af51d745-4c23-4595-bda3-161718e13132.png)

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+		};
+
+		const photoSection = {
+			cols: [
```
когда открываешь форму для нового контакта, там нету картинки по умолчанию
![Selection_999(863)](https://user-images.githubusercontent.com/31205423/166440608-a75a4853-a114-449d-9833-8374efa8f6a9.png)

kullias commented on sources/models/contacts.js:
```diff
@@ -1,11 +1,25 @@
 			data.value = `${data.FirstName} ${data.LastName}`;
+			data.StartDate = webix.Date.strToDate("%Y-%m-%d %H:%i")(data.StartDate);
+			data.Birthday = webix.Date.strToDate("%Y-%m-%d %H:%i")(data.Birthday);
+			data.Photo = data.Photo ? data.Photo : dummyPictureUrl;
```
я бы лучше не меняла в данных, а проверяла в самом темплейте и если нету - показывала дефолтную
так решится проблема с тем, что в форме для add contact там 404

vdmppv-xbsoftware replied:
я думал, что смогу так обработать, чтобы в темплейтах как раз не проверять
но уже вижу, что не получилось 
спасибо :)

kullias commented on sources/models/contacts.js:
```diff
@@ -1,11 +1,25 @@
+			if (typeof data.Birthday === "object") {
+				data.Birthday = webix.Date.dateToStr("%Y-%m-%d %H:%i")(data.Birthday);
+			}
+			data.Photo = data.Photo ? data.Photo : dummyPictureUrl;
```
это тогда тоже не надо будет

kullias commented on sources/views/activitiesTable.js:
```diff
@@ -0,0 +1,127 @@
+
+	filterTableByContact(id) {
+		if (id) {
+			this.$$(ACTIVITIES_DATATABLE_ID).filter("#ContactID#", this.contactId, true);
```
можно объект таблицы сохранить в init в проперти класса, часто она вызывается

kullias replied:
а, она уже сохранена. тогда надо тут использовать

vdmppv-xbsoftware replied:
да, видимо не заметил, спасибо

vdmppv-xbsoftware committed:
fixes

list items: fixed height.
add contact form: added default icon, unselected list

kullias commented on sources/styles/app.css:
```diff
@@ -57,22 +57,35 @@
+    display: flex;
+    white-space: normal;
+    overflow: visible !important;
+    height: auto !important;
```
вообще для сайзинга лучше использовать type:{ height:"auto" } в настройках листа. тут все норм и так, насколько я вижу. но 1) js коду вебиксе чаще всего все равно что там в css и могут быть проблемы с позиционированием, размерностью и скроллом 2) !important в стилях там, где без них можно обойтись - не очень

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+					width: 220,
+					height: 220,
+					borderless: true,
+					template: obj => `<img src="${obj.Photo ? obj.Photo : dummyPictureUrl}" class="contact-photo">`
```
obj.Photo || dummyPicture

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+							value: "Delete photo",
+							css: "webix_danger",
+							click: () => {
+								this.$$(CONTACT_PHOTO_ID).setValues({Photo: dummyPictureUrl});
```
Photo: "" будет правильнее

kullias replied:
по тому же подходу "не трогать данные, если там нету реальной картинки"

kullias commented on sources/views/contactsView.js:
```diff
@@ -1,12 +1,76 @@
+		this.on(this.app, "onContactSelect", (id) => {
+			this.list.unselectAll();
+
+			this.list.select(this.list.getFirstId());
```
тут по-любому что-то селектится, потому unselect не надо
ну а если селектить нечего - то это когда вообще ничего в листе нету (все удалили). потому надо проверка на это

```js
if (id)
  select(id);
else {
  const first = get first id
  if (first) select(first);
}
```

vdmppv-xbsoftware committed:
fixes [1]

kullias commented on sources/views/contactsInfo.js:
```diff
@@ -69,4 +144,24 @@ export default class ContactsInfo extends JetView {
+		webix.confirm({text: "Are you sure you want to delete this contact and all related files and activities"}).then(() => {
+			activitiesCollection.data.each((activity) => {
+				if (+activity.ContactID === +this.contactId) {
+					activitiesCollection.remove(activity.id);
```
как-то они некорректно удаляются:
- создаем контакт
- добавляем ему две активити
- удаляем контакт
- одна активити удалилась, вторая висит, но уже без контакта в столбце
- открываем форму этой активити - в селект подставляется другой контакт

kullias replied:
bug: https://snippet.webix.com/i2uzkpss
fixed: не удаляй во время прохода

vdmppv-xbsoftware replied:
В примере вижу, что удаляется лишь часть элементов (в данном случае нечётные)
Видимо в моём случае происходит так же

kullias replied:
если передать методу remove массив айдишек, он сам по нему пройдется и все удалит

kullias replied:
`if (activity) ` - а что, в массиве будут пустышки?

kullias commented on sources/views/contactsInfo.js:
```diff
@@ -69,4 +144,24 @@ export default class ContactsInfo extends JetView {
+
+			filesCollection.data.each((file) => {
+				if (+file.ContactID === +this.contactId) {
+					filesCollection.remove(file.id);
```
с файлами наверное то же самое (надо подебажить). но поскольку у нас нет общей таблицы файлов - так не заметно

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+			css: "form-header"
+		};
+
+		const firstCol = {
```
на маленьком экране печалити
![Selection_999(864)](https://user-images.githubusercontent.com/31205423/166452717-cb9aa54f-aab7-42d1-b18a-948289699e9a.png)

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+
+	urlChange() {
+		this.contactform = this.$$(CONTACTS_FORM_ID);
+		this.contactform.clear();
```
учитывая что форму открывает инфо/парент и при смене айди форма помирает, открывается инфо, оно не особо надо, скорее всего

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+		if (this.contactId) {
+			const item = contactsCollection.getItem(this.contactId);
+			this.contactform.setValues(item);
+			this.contactphoto.setValues({Photo: item.Photo});
```
норм вообще, но еще круче было бы сделать маленький кастом виджет, который работал бы как контрол в форме (у него должны быть методы setValue, getValue и ему можно было бы задать name и апи формы с ним бы работало)
но это просто скорее fyi

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,277 @@
+	cancelForm() {
+		this.app.callEvent("onContactSelect", [this.contactId]);
+
+		this.contactform.clear();
```
мне кажется и тут не надо, из-за того что селект листа вызывает в итоге закрытие формы, ее разрушение

kullias commented on sources/views/popupEditor.js:
```diff
@@ -112,9 +114,16 @@ export default class PopupEditor extends JetView {
+	showPopupEditor(id, contactId) {
+		if (!id && contactId) {
+			const currentContact = contactsCollection.getItem(contactId);
+			this.form.setValues({ContactID: currentContact.id});
```
а можно ж сразу contactId туда отправить, без getItem
или есть какие-то причины доставать объект и оттуда ту же айди брать?

vdmppv-xbsoftware committed:
small fix + layout fix

vdmppv-xbsoftware committed:
fixed contact deletion

vdmppv-xbsoftware committed:
fixes [2]

kullias commented on sources/views/contactsForm.js:
```diff
@@ -0,0 +1,269 @@
+					view: "text",
+					name: "Phone",
+					label: "Phone",
+					required: true
```
если делать таким полям валидацию, то правдоподобную лучше. щас можно ввести номер типа фыорпфырвп и он проканает

vdmppv-xbsoftware committed:
validation fix
